# Evaluation of retrieval using different parameters - search strategy

## What are we evaluating?
Based on the embedding model experiments in `04_evaluation-ret-exp.ipynb`, 
`pritamdeka/S-PubMedBert-MS-MARCO` was selected as the best embedding model.

Here we compare different search strategies:
- **Vector search** — baseline, semantic similarity only
- **Hybrid search** — vector + text search combined with RRF
- **Hybrid + query rewriting** — LLM rewrites query before hybrid search
- **Hybrid + reranking** — cross-encoder reranks hybrid search results

## Metrics
- **Hit Rate@10** — proportion of questions where correct chunk appears in top 10
- **MRR@10** — Mean Reciprocal Rank

### Step 1. Connect to the database

In [ ]:
import psycopg
import pandas as pd
from dotenv import load_dotenv
from openai import OpenAI
from sentence_transformers import SentenceTransformer, CrossEncoder
from app.ragbase import RAGBaseEval
from pipelines.evaluation import evaluate

load_dotenv()
openai_client = OpenAI()

conn = psycopg.connect("postgresql://user:pswd@localhost:5432/papers")

c:\Users\marta\Documents\projects\literature-assistant-llm\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### Step 2. Load ground truth

In [ ]:
df_question = pd.read_csv("../data/ground_truth.csv")
ground_truth_retrieval = df_question.to_dict(orient="records")

In [5]:
### Step 3. Evaluate

In [8]:
embedding_model = SentenceTransformer("pritamdeka/S-PubMedBert-MS-MARCO")
reranker = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")


# hybrid search
class RAGHybrid(RAGBaseEval):
    def search(self, query, num_results=10):
        return self.hybrid_search(query, num_results=num_results)

# hybrid search + query rewriting
class RAGHybridRewrite(RAGBaseEval):
    def search(self, query, num_results=10):
        rewritten = self.rewrite_query(query)
        return self.hybrid_search(rewritten, num_results=num_results)


# hybrid search + reranking

class RAGHybridRerank(RAGBaseEval):
    def search(self, query, num_results=10):
        # get more results first then rerank
        results = self.hybrid_search(query, num_results=20)
        
        # rerank
        pairs = [(query, doc["text"]) for doc in results]
        scores = reranker.predict(pairs)
        
        # sort by reranker score
        reranked = sorted(zip(scores, results), key=lambda x: x[0], reverse=True)
        
        return [doc for _, doc in reranked[:num_results]]
    
# evaluate all three
approaches = [
    {"name": "hybrid", "class": RAGHybrid},
    {"name": "hybrid + rewrite", "class": RAGHybridRewrite},
    {"name": "hybrid + rerank", "class": RAGHybridRerank}
]

results = []
for approach in approaches:
    print(f"\nEvaluating: {approach['name']}")
    assistant = approach["class"](
        conn=conn,
        model=embedding_model,
        llm_client=openai_client,
        table_name="chunks"
    )
    metrics = evaluate(ground_truth_retrieval, assistant.search)
    results.append({
        "approach": approach["name"],
        "hit_rate": metrics["hit_rate"],
        "mrr": metrics["mrr"]
    })


import pandas as pd
df = pd.DataFrame(results)
print(df)

Loading weights: 100%|██████████| 105/105 [00:00<00:00, 4374.14it/s]



Evaluating: hybrid


100%|██████████| 7430/7430 [48:47<00:00,  2.54it/s]  



Evaluating: hybrid + rewrite


100%|██████████| 7430/7430 [3:03:53<00:00,  1.49s/it]  



Evaluating: hybrid + rerank


100%|██████████| 7430/7430 [3:44:53<00:00,  1.82s/it]  

           approach  hit_rate       mrr
0            hybrid  0.632571  0.417047
1  hybrid + rewrite  0.430552  0.246761
2   hybrid + rerank  0.691521  0.555747


## Conclusion

Based on the retrieval evaluation results, **hybrid search with cross-encoder reranking** was selected as the final retrieval pipeline:

| Search strategy | Hit Rate@10 | MRR@10 |
|----------------|-------------|--------|
| Hybrid search | 0.633 | 0.417 |
| Hybrid + query rewriting | 0.431 | 0.247 |
| **Hybrid + reranking** | **0.692** | **0.556** |

**Key findings:**

- **Hybrid search** outperforms pure vector search by combining semantic similarity (vector) with exact keyword matching (BM25 text search), merged with Reciprocal Rank Fusion (RRF)
- **Query rewriting hurt performance** — rewriting scientific questions with the LLM lost specificity and reduced retrieval quality
- **Cross-encoder reranking** (`cross-encoder/ms-marco-MiniLM-L-6-v2`) significantly improved both Hit Rate (+9%) and MRR (+33%) over plain hybrid search by reranking the top 20 candidates

**Selected configuration:**
- Hybrid search (vector + BM25) retrieving top 20 candidates
- Cross-encoder reranking returning top 10 results
- Embedding model: `pritamdeka/S-PubMedBert-MS-MARCO`

This configuration is used in the final RAG pipeline (`RAGBaseHistory` in `app/ragbase.py`).